quiz/quiz.go

```diff
@@ -0,0 +1,165 @@
+// stdin with "\n" being used to specify the end of an answer.
+func (q *Quiz) Start(maxQuizTime time.Duration, random bool) error {
+	fmt.Printf("\nYou will have %v seconds to answer %v questions.\nPress [Enter] to Start the quiz.\n",
+		maxQuizTime, q.TotalQuestions)
+
+	reader := bufio.NewReader(os.Stdin)
+	_, err := reader.ReadString('\n')
+	if err != nil {
+		return err
+	}
+
+	completed := make(chan string)
+
+	if random {
+		go proctorRandomQuiz(q, reader, completed)
+	} else {
+		go proctorQuiz(q, reader, completed)
+	}
+
+	timeout := time.NewTimer(maxQuizTime)
+	defer timeout.Stop()
+
+	select {
+	case <-timeout.C:
+		return fmt.Errorf("the quiz timer (%v seconds) ran out before the user completed the quiz",
+			maxQuizTime)
+	case <-completed:
+		return nil
+	}
+}
+
+// proctorQuiz performs the quiz, asking the user questions and gathering responses. When it finishes, it sends
+// a "success" message down the channel to be returned to the calling function
+func proctorQuiz(q *Quiz, reader *bufio.Reader, c chan string) {
+	for index, value := range q.Questions {
+		processQuestion(reader, &value, q, index)
+	}
+	c <- "done"
+}
+
+func proctorRandomQuiz(q *Quiz, reader *bufio.Reader, c chan string) {
+	questionNumber := 1
+	for len(q.Questions) > 0 {
+		// gather a random index to get a random question
+		randomIndex := rand.Intn(len(q.Questions))
+
+		// store the random question
+		randomQuestion := q.Questions[randomIndex]
+
+		// remove the random question from the Questions slice
+		frontSlice := q.Questions[0:randomIndex]
+		backSlice := q.Questions[randomIndex+1 : len(q.Questions)]
+		q.Questions = append(frontSlice, backSlice...)
+
+		processQuestion(reader, &randomQuestion, q, questionNumber)
+		questionNumber++
+	}
+	c <- "done"
+}
+
+func processQuestion(reader *bufio.Reader, question *Question, quiz *Quiz, questionNumber int) {
+	if question.QuestionText != "" {
+		fmt.Print("\n\tQuestion ", questionNumber, ":", "\t\t"+question.QuestionText+" = ")
+
+		userAnswer, inputErr := reader.ReadString('\n')
+		if inputErr == nil {
+			userAnswer = strings.Replace(userAnswer, "\n", "", 1)
+		} else {
+			fmt.Println("An error has occurred reading user input: ", inputErr)
+		}
+
+		userAnswer = strings.TrimSpace(userAnswer)
+		userAnswer = strings.ToLower(userAnswer)
+		result := userAnswer == strings.ToLower(question.Answer)
+		if result {
+			fmt.Println("\tCorrect!")
+			quiz.CorrectQuestions++
+		} else {
+			fmt.Println("\tIncorrect.")
+		}
+		question.Correct = &result
+		question.UserAnswer = userAnswer
+	}
+}
+
+// Results takes a Quiz and performs the processing needed to determine the % of correct answers
+func (q Quiz) Results() string {
+	correct, total := float64(q.CorrectQuestions), float64(q.TotalQuestions)
+	ratio := 100 * (correct / total)
+	percent := "%"
+	return fmt.Sprintf("%2.2f%v (%v/%v) of the answers were correct.", ratio, percent, correct, total)
```

Comment:
What is going on with this string? `%` characters are represented by `%%` in fmt statements. Don't declare a extra value to work around that.